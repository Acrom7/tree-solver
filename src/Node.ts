export enum NodeType {
    Number = 'Number',
    Operator = 'Operator',
    Empty = 'Empty'
}

export class Node {
    value: string | number
    left: Node | null
    right: Node | null

    constructor(nodeValue: string | number = ' ') {
        this.value = nodeValue
        this.left = null
        this.right = null
    }
}
