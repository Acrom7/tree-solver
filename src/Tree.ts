import {Node, NodeType} from './Node'
import {operators} from './types'

export class Tree {
    expression: string
    tokens: string[]
    root: Node

    constructor(expression: string) {
        this.expression = expression
        this.tokens = expression.split('').filter(el => el !== ' ')
        this.root = new Node()
        this.gen(this.tokens, this.root)
    }

    gen(tokens: string[], root: Node): void {
        if (tokens.length === 1) {
            root.value = tokens[0]
            root.type = NodeType.Number
            return
        }
        const lowPriorityIndex = this.getLowPriorityToken(tokens)
        // Если нет оператора вне скобок
        if (lowPriorityIndex === -1) {
            this.gen(tokens.slice(1, -1), root)
            return
        }
        root.value = tokens[lowPriorityIndex]
        root.type = NodeType.Operator
        root.left = new Node()
        this.gen(tokens.slice(0, lowPriorityIndex), root.left)
        root.right = new Node()
        this.gen(tokens.slice(lowPriorityIndex + 1), root.right)
    }

    // Возвращает приоритет оператора: чем больше возращаемое значение - тем больше приоритет
    getPriority = (operator: string): number => {
        // Сгруппированы в порядке возрастания приоритета
        const operatorsPriority = [['+', '-'], ['*', '/']]
        for (let i = 0; i < operatorsPriority.length; ++i) {
            if (operatorsPriority[i].includes(operator)) {
                return i
            }
        }
        return -1
    }

    // Возвращает индекс оператора с наименьшим приоритетом
    getLowPriorityToken = (tokensArray: string[]): number => {
        let lowPriority = Number.MAX_VALUE
        let index = -1
        let parenthesesCount = 0
        for (let i = 0; i < tokensArray.length; ++i) {
            const token = tokensArray[i]
            if (operators.includes(token) && parenthesesCount === 0) {
                const tokenPriority = this.getPriority(token)
                if (tokenPriority < lowPriority) {
                    lowPriority = tokenPriority
                    index = i
                }
            } else if (token === '(') {
                ++parenthesesCount
            } else if (token === ')') {
                --parenthesesCount
            }
        }
        return index
    }

    print() {
        // tslint:disable
        // @ts-ignore
        function printNode(printFn, node, leftKey, rightKey, displayKey, indent) {
            // default to empty string to avoid logging the word 'undefined'
            indent = indent || ''

            // current item's indentation prefix
            // tslint:disable-next-line:prefer-const
            var prefix = indent + '\\-'

            var leftChild = leftKey.constructor === Function ? leftKey(node) : node[leftKey],
                rightChild = rightKey.constructor === Function ? rightKey(node) : node[rightKey]

            var displayStr = displayKey.constructor === Function ? displayKey(node) : node[displayKey]
            printFn(prefix + ' ' + displayStr)

            // recurse left
            if (!!leftChild) printNode(printFn, leftChild, leftKey, rightKey, displayKey, indent + (!!rightChild ? ' |' : '  '))
            // recurse right
            if (!!rightChild) printNode(printFn, rightChild, leftKey, rightKey, displayKey, indent + '  ')
        }

        // tslint:enable
        printNode(console.log, this.root, 'left', 'right', 'value', undefined)
    }
}
