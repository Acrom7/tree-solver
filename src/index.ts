import {Tree} from './Tree'

const expr = '1 + 1'
const tree = new Tree()
tree.expr(expr)
tree.print()
console.log(tree.solve())
