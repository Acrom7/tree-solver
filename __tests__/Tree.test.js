import {Tree} from "../src/Tree"

describe('Mathematical expressions', () => {
	test.each([
		['1 + 1', 2],
		['2 + 2 * 2', 6],
		['(1 + 2) * ( 3 - 4 ) * 9', -27],
		['((1 + 1) / 2 - (5 * (3 + 1)))', -19],
		['1 / 0', Infinity],
	])('%s === %d', (expr, result) => {
		const tree = new Tree()
		tree.expr(expr)
		expect(tree.solve()).toBe(result)
	})
})

describe('Inserting nodes', () => {
	let tree

	beforeEach(() => {
		tree = new Tree()
	})

	it('should be simple and return 2', () => {
		tree.insert('+', 1, 1)
		tree.insert('1', 2, 1)
		tree.insert(1, 2, 2)
		expect(tree.solve()).toBe(2)
		expect(tree).toMatchSnapshot()
	})

	it('should throw error on empty tree', () => {
		expect(() => tree.solve()).toThrowError(/empty/i)
		expect(tree).toMatchSnapshot()
	})

	it('should parse expression, insert some nodes, and return 88', () => {
		tree.expr('1 + 2 * (3 + 4)')
		// replace 1 on 11
		tree.insert(11, 2, 1)
		// replace 2 on +, add arguments 5 and 6
		tree.insert('+', 3, 3)
		tree.insert(5, 4, 5)
		tree.insert(6, 4, 6)
		expect(tree.solve()).toBe(88)
		expect(tree).toMatchSnapshot()
	})

	it('should replace root node operator', () => {
		tree.insert('+', 1, 1)
		tree.insert('1', 2, 1)
		tree.insert(1, 2, 2)
		tree.insert('-', 1, 1)
		expect(tree.solve()).toBe(0)
		expect(tree).toMatchSnapshot()
	})

	it('should not insert value into level 2 on empty tree', () => {
		const result = tree.insert('1', 2, 1)
		expect(result).toBe(-1)
		expect(tree).toMatchSnapshot()
	})
})
