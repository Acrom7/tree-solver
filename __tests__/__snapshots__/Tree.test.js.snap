// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Inserting nodes should be simple and return 2 1`] = `
Tree {
  "getLowPriorityToken": [Function],
  "getPriority": [Function],
  "root": Node {
    "left": Node {
      "left": null,
      "right": null,
      "value": "1",
    },
    "right": Node {
      "left": null,
      "right": null,
      "value": 1,
    },
    "value": "+",
  },
}
`;

exports[`Inserting nodes should not insert value into level 2 on empty tree 1`] = `
Tree {
  "getLowPriorityToken": [Function],
  "getPriority": [Function],
  "root": null,
}
`;

exports[`Inserting nodes should parse expression, insert some nodes, and return 88 1`] = `
Tree {
  "getLowPriorityToken": [Function],
  "getPriority": [Function],
  "root": Node {
    "left": Node {
      "left": null,
      "right": null,
      "value": 11,
    },
    "right": Node {
      "left": Node {
        "left": Node {
          "left": null,
          "right": null,
          "value": 5,
        },
        "right": Node {
          "left": null,
          "right": null,
          "value": 6,
        },
        "value": "+",
      },
      "right": Node {
        "left": Node {
          "left": null,
          "right": null,
          "value": "3",
        },
        "right": Node {
          "left": null,
          "right": null,
          "value": "4",
        },
        "value": "+",
      },
      "value": "*",
    },
    "value": "+",
  },
}
`;

exports[`Inserting nodes should replace root node operator 1`] = `
Tree {
  "getLowPriorityToken": [Function],
  "getPriority": [Function],
  "root": Node {
    "left": Node {
      "left": null,
      "right": null,
      "value": "1",
    },
    "right": Node {
      "left": null,
      "right": null,
      "value": 1,
    },
    "value": "-",
  },
}
`;

exports[`Inserting nodes should throw error on empty tree 1`] = `
Tree {
  "getLowPriorityToken": [Function],
  "getPriority": [Function],
  "root": null,
}
`;
